src/internal/pfsdb/repos.go

```diff
@@ -0,0 +1,363 @@
+	if !ok {
+		return false
+	}
+	return targetErr.Code == "23505" // duplicate key SQLSTATE
+}
+
+// RepoIterator batches a page of repoRow entries. Entries can be retrieved using iter.Next().
+type RepoIterator struct {
+	limit    int
+	offset   int
+	repos    []repoRow
+	index    int
+	tx       *pachsql.Tx
+	where    string
+	whereVal interface{}
+}
+
+type repoRow struct {
+	ID          pachsql.ID `db:"id"`
+	Name        string     `db:"name"`
+	ProjectID   string     `db:"project_id"`
+	ProjectName string     `db:"proj_name"`
+	Description string     `db:"description"`
+	RepoType    string     `db:"type"`
+	CreatedAt   time.Time  `db:"created_at"`
+	UpdatedAt   time.Time  `db:"updated_at"`
+	// Branches is a string that contains an array of hex-encoded branchInfos. The array is enclosed with curly braces.
+	// Each entry is prefixed with '//x' and entries are delimited by a ','
+	Branches string `db:"branches"`
+}
+
+// NextWithID is like Next, but populates the 'id' and 'dst' parameters each iteration where 'id' is a pointer to the row ID and dst is the repoInfo.
+func (iter *RepoIterator) NextWithID(ctx context.Context, id *pachsql.ID, dst **pfs.RepoInfo) error {
+	if dst == nil {
+		return errors.Wrap(fmt.Errorf("repo is nil"), "get next repo")
+	}
+	var err error
+	if iter.index >= len(iter.repos) {
+		iter.index = 0
+		iter.offset += iter.limit
+		iter.repos, err = listRepoPage(ctx, iter.tx, iter.limit, iter.offset, iter.where, iter.whereVal)
+		if err != nil {
+			return errors.Wrap(err, "list repo page")
+		}
+		if len(iter.repos) == 0 {
+			return stream.EOS()
+		}
+	}
+	row := iter.repos[iter.index]
+	*dst, err = getRepoFromRepoRow(&row)
+	if id != nil {
+		*id = pachsql.ID(row.ID)
+	}
+	if err != nil {
+		return errors.Wrap(err, "getting repoInfo from repo row")
+	}
+	iter.index++
+	return nil
+}
+
+// Next advances the iterator by one row. It returns a stream.EOS when there are no more entries.
+func (iter *RepoIterator) Next(ctx context.Context, dst **pfs.RepoInfo) error {
+	return iter.NextWithID(ctx, nil, dst)
+}
+
+// ListRepo returns a RepoIterator that exposes a Next() function for retrieving *pfs.RepoInfo references.
+func ListRepo(ctx context.Context, tx *pachsql.Tx) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "", nil)
+}
+
+// ListRepoByIdxType is like ListRepo but only iterates over repo.type = repoType.
+func ListRepoByIdxType(ctx context.Context, tx *pachsql.Tx, repoType string) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "type", repoType)
+}
+
+// ListRepoByIdxName is like ListRepo but only iterates over repo.name = name.
+func ListRepoByIdxName(ctx context.Context, tx *pachsql.Tx, repoName string) (*RepoIterator, error) {
+	return listRepo(ctx, tx, "name", repoName)
+}
+
+// ListRepo returns a RepoIterator that exposes a Next() function for retrieving *pfs.RepoInfo references.
+func listRepo(ctx context.Context, tx *pachsql.Tx, where string, whereVal interface{}) (*RepoIterator, error) {
```

Comment:
This `whereVal` is a little suspicious, but I think it's fine for now.

Reply:
You're right, all the callers are string types, so it doesn't need to be an interface. I think for projects there was a whereVal pattern where one caller was an int type and the other was a string type, but that is not needed here. Fixed!